### PR TITLE
reimplemented the MemQueue class to avoid unordered inserts

### PR DESCRIPTION
**Pull request by tarekziade**
I closed https://github.com/elastic/connectors-python/pull/281 because I figured out that we did not have a race condition.

See https://github.com/elastic/connectors-python/pull/281#issuecomment-1379794960

But then I realized the implementation can lead to unordered inserts in the queue. So this new patch reintroduces the change which does respect the ordering.

### CODE REVIEW

**Review by pquentin — APPROVED** [review]
Thanks!

File: connectors/tests/test_utils.py [hunk 1]
```diff
@@ -51,13 +53,65 @@ def test_invalid_names():
...
+    # vanilla queue
+    queue = asyncio.Queue()
+    start = time.time()
+    for i in range(1000):
+        await queue.put("x" * 100)
+    queue_duration = time.time() - start
+
+    # mem queue should be 30 times slower at the most
+    quotient, _ = divmod(mem_queue_duration, queue_duration)
+    assert quotient < 30
```
Comment (artem-shelkovnikov) [inline connectors/tests/test_utils.py:74]: Why 30? :)

File: connectors/tests/test_utils.py [hunk 2]
```diff
@@ -51,13 +53,65 @@ def test_invalid_names():
...
+
+
+@pytest.mark.asyncio
+async def test_mem_queue_race(patch_logger):
+    item = "small stuff"
+    queue = MemQueue(
+        maxmemsize=get_size(item) * 2 + 1, refresh_interval=0.1, refresh_timeout=30
+    )
+    max_size = 0
```
Comment (artem-shelkovnikov) [inline connectors/tests/test_utils.py:83]: I think a comment would help to understand what's going on in the test. I vaguely understand it, but not 100% sure and not 100% sure what is the actual problem.

**Review by artem-shelkovnikov — COMMENTED** [review]
I tried to review, but it's still pretty hard to understand what's going on. Is it possible to add more tests/comments or refactor to make the logic easier to read?

File: connectors/tests/test_utils.py (hunk 1 above)
Comment (tarekziade) [reply connectors/tests/test_utils.py:74]: Arbitrary number. It's just set to make sure the class does not get slower over time when we change it

**Comment by tarekziade** [comment]
> I tried to review, but it's still pretty hard to understand what's going on. Is it possible to add more tests/comments or refactor to make the logic easier to read?

I can add comments, but hardly refactor because it's based on the `asyncio.Queue` class logic

**Comment by tarekziade** [comment]
> > I can add comments, but hardly refactor because it's based on the asyncio.Queue class logic
> 
> I see. Maybe pointing out every change to the original implementation can help?

this is what I did with the two comments (1/ 2/) in the cod, we have two changes -- the original method is here https://github.com/python/cpython/blob/3.10/Lib/asyncio/queues.py#L111

**Review by artem-shelkovnikov — APPROVED** [review]
LGTM

**Pull request merged**